//ZE_SOURCE_PROCESSOR_START(License, 1.0)
/*******************************************************************************
 Zinek Engine - ZEResourceConfigurationWidget.cpp
 ------------------------------------------------------------------------------
 Copyright (C) 2008-2021 Yiğit Orçun GÖKBULUT. All rights reserved.

 This file is part of the Zinek Engine  Software. Zinek Engine Software and the
 accompanying  materials are  made available  under the  terms of Zinek  Engine
 Commercial License or the GNU General Public License Version 3.

                      ZINEK ENGINE COMMERCIAL LICENSE
 Licensees  holding  valid  Zinek Engine Commercial  License(s) may  use  Zinek
 Engine  Software in  accordance  with   the  commercial  license  agreement(s)
 (which can only be  issued  by  copyright  owner "Yiğit  Orçun  GÖKBULUT") and
 provided with the Software  or, alternatively,  in  accordance with the  terms
 contained  in  a  written  agreement  between  you  and  copyright  owner. For
 licensing  terms  and conditions  please  contact  with  copyright owner.

                    GNU GENERAL PUBLIC LICENSE VERSION 3
 This program is free software: you can  redistribute it and/or modify it under
 the terms of the GNU General Public  License as published by the Free Software
 Foundation, either  version 3 of  the License, or  (at your option)  any later
 version. This program is  distributed in the hope that  it will be useful, but
 WITHOUT ANY WARRANTY; without even the  implied warranty of MERCHANTABILITY or
 FITNESS FOR A PARTICULAR PURPOSE. See  the GNU General Public License for more
 details. You  should have received  a copy of the  GNU General  Public License
 along with this program. If not, see <https://www.gnu.org/licenses/>.

 Copyright Owner Information:
  Name: Yiğit Orçun GÖKBULUT
  Contact: orcun.gokbulut@gmail.com
  Github: https://www.github.com/orcun-gokbulut/ZE
*******************************************************************************/
//ZE_SOURCE_PROCESSOR_END()

#include "ZEResourceConfigurationWidget.h"
#include "ui_ZEResourceConfigurationWidget.h"
#include "ZEResourceOptionTreeItem.h"
#include "QtGui/QTreeWidgetItem"
#include "ZEResourceOption.h"
#include "ZEML/ZEMLNode.h"
#include "ZEFile/ZEFile.h"
#include "ZEML/ZEMLProperty.h"

ZEResourceConfigurationWidget::ZEResourceConfigurationWidget(QWidget* Parent)
{
	QApplicationCreated = false;

	if(QApplication::instance() == NULL)
	{
		ZEInt32	Argc = 0;
		void*	Argv = NULL;
		Application = new QApplication(Argc, NULL);
		QApplicationCreated = true;
	}

	Form = new Ui::ZEResourceConfiguratorUI();
	Dialog = new QDialog(Parent);
	Form->setupUi(Dialog);
	Dialog->setWindowTitle("Resource Configuration");

	OptionWidget = new ZEResourceOptionWidget(Form->OptionsParentWidget);
	Form->OptionsParentWidgetLayout->addWidget(OptionWidget);
	OptionWidget->setVisible(false);
	SignalHandler = new ZEResourceConfigurationWidgetSignalHandler(this, Dialog);

	Preset = NULL;
}

ZEResourceConfigurationWidget::~ZEResourceConfigurationWidget()
{
	for (ZESize I = 0; I < ResourceOptions.GetCount(); I++)
	{
		delete ResourceOptions[I];
		ResourceOptions[I] = NULL;
	}

	Dialog->hide();
	delete Dialog;
	Dialog = NULL;

	delete Form;
	Form = NULL;

	if(Preset != NULL)
		delete Preset;

	if(QApplicationCreated)
	{
		Application->quit();
		delete Application;
		Application = NULL;
	}
}

void ZEResourceConfigurationWidget::AddResource(const ZEString& Identifier, const ZEString& PhysicalPath, const ZEString& ExportPath, ZEResourceOptionAvailableAction Action, bool IsAutoGenerated)
{
	ZEArray<ZEMLNode*> SubNodes;

	if(Preset != NULL)
		SubNodes = Preset->GetSubNodes(Identifier);

	ZEResourceOption* NewOption = new ZEResourceOption();
	ResourceOptions.Add(NewOption);
	ZEResourceOptionTreeItem* NewItem = new ZEResourceOptionTreeItem();

	ResourceOptions.GetLastItem()->Identifier = Identifier;

	if(SubNodes.GetCount() == 0)
	{
		ResourceOptions.GetLastItem()->PhysicalPath = PhysicalPath;
		ResourceOptions.GetLastItem()->ExportPath = ExportPath;
		ResourceOptions.GetLastItem()->Action = Action;
		ResourceOptions.GetLastItem()->IsAutoGenerated = IsAutoGenerated;
	}
	else if(SubNodes.GetCount() > 0)
	{
		if(SubNodes.GetCount() != 1)
			zeWarning("Duplicate resource identifier found.");

		for (ZESize I = 0; I < SubNodes[0]->GetProperties().GetCount(); I++)
		{
			if(SubNodes[0]->GetProperties()[I]->GetName() == ZEString("ExportPath"))
				ResourceOptions.GetLastItem()->ExportPath = ((ZEMLProperty*)(SubNodes[0]->GetProperties()[I]))->GetValue().GetString();

			else if(SubNodes[0]->GetProperties()[I]->GetName() == ZEString("PhysicalPath"))
				ResourceOptions.GetLastItem()->PhysicalPath = ((ZEMLProperty*)(SubNodes[0]->GetProperties()[I]))->GetValue().GetString();

			else if(SubNodes[0]->GetProperties()[I]->GetName() == ZEString("Action"))
				ResourceOptions.GetLastItem()->Action = (ZEResourceOptionAvailableAction)((ZEMLProperty*)(SubNodes[0]->GetProperties()[I]))->GetValue().GetInt32();

			else if(SubNodes[0]->GetProperties()[I]->GetName() == ZEString("IsAutoGenerated"))
				ResourceOptions.GetLastItem()->IsAutoGenerated = ((ZEMLProperty*)(SubNodes[0]->GetProperties()[I]))->GetValue().GetBoolean();
		}
	}

	NewItem->SetResourceOption(ResourceOptions[ResourceOptions.GetCount() - 1]);
	Form->ResourcesTreeWidget->addTopLevelItem(NewItem);
	NewItem->Update();
}

void ZEResourceConfigurationWidget::AddResource(const ZEString& Identifier, const ZEString& ExportPath, ZEResourceOptionAvailableAction Action, bool IsAutoGenerated)
{
	AddResource(Identifier ,ZEString(), ExportPath, Action, IsAutoGenerated);
}

bool ZEResourceConfigurationWidget::Show()
{
	if(Dialog->exec() == QDialog::Accepted)
		return true;
	else
		return false;
}

void ZEResourceConfigurationWidget::Hide()
{
	Dialog->close();
}

const ZEArray<ZEResourceOption*>& ZEResourceConfigurationWidget::GetOptions()
{
	return ResourceOptions;
}

void ZEResourceConfigurationWidget::SetPresetFilePath(const ZEString& Path)
{
	if(Preset != NULL)
	{
		delete Preset;
		Preset = NULL;
	}

	ZEFile PresetFile;
	
	if(!PresetFile.Open(Path, ZE_FOM_READ, ZE_FCM_NONE))
		zeWarning("Preset file not found.");
	else
	{
		Preset = new ZEMLNode("Preset");
		Preset->Read(&PresetFile);
		PresetFile.Close();
	}
}

void ZEResourceConfigurationWidget::SavePresets(const ZEString& Path)
{
	ZEFile PresetFile;

	if(!PresetFile.Open(Path, ZE_FOM_READ_WRITE, ZE_FCM_OVERWRITE))
		zeError("Cannot open preset file for saving presets %s.", Path);
	else
	{
		if(Preset != NULL)
		{
			delete Preset;
			Preset = NULL;
		}
		
		if(Preset == NULL)
			Preset = new ZEMLNode("Preset");

		for(ZESize I = 0; I < ResourceOptions.GetCount(); I++)
		{
			ZEMLNode* Node = Preset->AddSubNode(ResourceOptions[I]->Identifier);
			Node->AddProperty("ExportPath", ResourceOptions[I]->ExportPath);
			Node->AddProperty("PhysicalPath", ResourceOptions[I]->PhysicalPath);
			Node->AddProperty("Action", ResourceOptions[I]->Action);
			Node->AddProperty("IsAutoGenerated", ResourceOptions[I]->IsAutoGenerated);
		}

		Preset->Write(&PresetFile);
		PresetFile.Close();
	}
}

const ZEResourceOption* ZEResourceConfigurationWidget::GetOption(const ZEString& Identifier)
{
	for (ZESize I = 0; I < ResourceOptions.GetCount(); I++)
	{
		if(ResourceOptions[I]->Identifier == Identifier)
			return ResourceOptions[I];
	}

	return NULL;
}
